 #### Instrucciones

Preparación del Entorno
Asegúrate de tener instalado Pandas en tu entorno de trabajo.
Descarga el archivo dataset.csv desde Kaggle. Elige un dataset que te interese y que no incluya visualización de datos. Algunas sugerencias pueden ser datasets relacionados con ventas, compras, productos, etc.



#### 1.- Preparación del entorno

In [111]:
# Se usa biblioteca google
from google.colab import drive
# Se conecta colab con drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [112]:
# Se importa librerias
import pandas as pd

#### 2.- Cargar los datos

Nota: Se elige de Kaggle.com Dataset 'ventas_starbuck_2025', la cual son
       ventas ficticias de distintas sucursales de tiendas Starbuck en Argentina.
* la base de datos se subio en https://github.com/danteginos/Mis_Proyectos/tree/main/DATA

In [113]:
# Se importa el DataFrame
path='/content/drive/MyDrive/DATA/ventas_starbucks_2025 (1).csv'
df = pd.read_csv(path)


In [114]:
df.head(10)

ID_Venta       Fecha   Hora                       Sucursal  \
0  VTA00001  2025-01-04  08:25  San Lorenzo 25, Nueva Córdoba   
1  VTA00002  2025-01-10  14:08               Córdoba Shopping   
2  VTA00003  2025-01-17  15:10               Córdoba Shopping   
3  VTA00004  2025-03-23  07:23           Calle San Lorenzo 47   
4  VTA00005  2025-03-19  18:33               Córdoba Shopping   
5  VTA00006  2025-02-02  18:10           Calle San Lorenzo 47   
6  VTA00007  2025-01-09  11:58           Calle San Lorenzo 47   
7  VTA00008  2025-02-20  14:28               Córdoba Shopping   
8  VTA00009  2025-02-25  13:53  San Lorenzo 25, Nueva Córdoba   
9  VTA00010  2025-02-23  13:12  San Lorenzo 25, Nueva Córdoba   

                            Nombre_Producto        Categoría_Producto  Tamaño  \
0                        Termo Reutilizable             Merchandising     NaN   
1                        Té Verde en Hebras                        Té     NaN   
2                       Moneda de Chocolate                    Snacks     NaN   
3                            Espresso Roast            Café en Granos     NaN   
4                            Espresso Roast            Café en Granos     NaN   
5  Croissant Relleno con Crema de Avellanas                    Bakery     NaN   
6                             Té Chai Latte               Té Caliente     NaN   
7                                     Mocha  Bebida Espresso Caliente  Grande   
8                            Espresso Roast            Café en Granos     NaN   
9                       Caramel Frappuccino               Frappuccino   Venti   

   Cantidad  Precio_Unitario  Total_Venta  ... Método_de_Pago Cliente_Miembro  \
0         2          3277.87      6555.74  ...            App              Sí   
1         1          1630.18      1630.18  ...        Tarjeta              Sí   
2         4           879.24      3516.96  ...        Tarjeta              No   
3         5          2230.73     11153.65  ...        Tarjeta              Sí   
4         5          2486.63     12433.15  ...        Tarjeta              No   
5         4           981.80      3927.20  ...            App              Sí   
6         2          1399.43      2798.86  ...       Efectivo              No   
7         3          1740.24      5220.72  ...       Efectivo              Sí   
8         4          2519.29     10077.16  ...            App              No   
9         4          1915.53      7662.12  ...            App              Sí   

  Descuento_Miembro  Promoción_Aplicada Empleado_Atendió   Turno  \
0                15               Combo           Camila   Noche   
1                10             Ninguna        Florencia   Noche   
2                 0             Ninguna           Julián   Tarde   
3                10             Ninguna           Marcos   Noche   
4                 0               Combo            Sofía   Tarde   
5                15          Happy Hour           Marcos   Noche   
6                 0             Ninguna            Sofía  Mañana   
7                15          Happy Hour           Camila   Noche   
8                 0               Combo           Marcos   Noche   
9                15                 2x1             Luis  Mañana   

  Demora_Preparación_Min  Satisfacción_Cliente  Stock_Antes  Stock_Después  
0                   13.7                     5           23             21  
1                    2.3                     4           20             19  
2                    8.8                     2           23             19  
3                    3.9                     3           19             14  
4                   13.8                     5           41             36  
5                    4.4                     4           21             17  
6                    5.8                     3           38             36  
7                    8.9                     5           40             37  
8                    5.0                     1           29  

#### 3.- Exploración Inicial de los Datos

In [115]:
df.tail(5)

ID_Venta       Fecha   Hora                       Sucursal  \
4995  VTA04996  2025-03-06  08:07              Avenida Colón 608   
4996  VTA04997  2025-01-17  17:46              Avenida Colón 608   
4997  VTA04998  2025-02-06  10:18  San Lorenzo 25, Nueva Córdoba   
4998  VTA04999  2025-03-12  11:10               Córdoba Shopping   
4999  VTA05000  2025-03-08  15:54              Avenida Colón 608   

                 Nombre_Producto      Categoría_Producto Tamaño  Cantidad  \
4995  Fosforito de Jamón y Queso              Sándwiches    NaN         2   
4996         Moneda de Chocolate                  Snacks    NaN         2   
4997        Croissant de Manteca                  Bakery    NaN         2   
4998   Wrap de Pollo y Vegetales              Sándwiches    NaN         5   
4999                Latte Helado  Bebida Espresso Helada  Venti         5   

      Precio_Unitario  Total_Venta  ... Método_de_Pago Cliente_Miembro  \
4995          1694.30      3388.60  ...            App              Sí   
4996          1106.72      2213.44  ...        Tarjeta              Sí   
4997          1430.26      2860.52  ...        Tarjeta              No   
4998          1524.23      7621.15  ...        Tarjeta              No   
4999          1648.32      8241.60  ...       Efectivo              No   

     Descuento_Miembro  Promoción_Aplicada Empleado_Atendió   Turno  \
4995                 0          Happy Hour           Julián   Noche   
4996                10             Ninguna           Camila  Mañana   
4997                 0             Ninguna           Marcos  Mañana   
4998                 0          Happy Hour        Florencia  Mañana   
4999                 0                 2x1        Florencia   Tarde   

     Demora_Preparación_Min  Satisfacción_Cliente  Stock_Antes  Stock_Después  
4995                   10.9                     2           41             39  
4996                    4.7                     4           26             24  
4997                   11.6                     1           30             28  
4998                    6.0                     1           17             12  
4999                   13.7                     5           39             34  

[5 rows x 21 columns]

In [116]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ID_Venta                5000 non-null   object 
 1   Fecha                   5000 non-null   object 
 2   Hora                    5000 non-null   object 
 3   Sucursal                5000 non-null   object 
 4   Nombre_Producto         5000 non-null   object 
 5   Categoría_Producto      5000 non-null   object 
 6   Tamaño                  1370 non-null   object 
 7   Cantidad                5000 non-null   int64  
 8   Precio_Unitario         5000 non-null   float64
 9   Total_Venta             5000 non-null   float64
 10  Método_de_Compra        5000 non-null   object 
 11  Método_de_Pago          5000 non-null   object 
 12  Cliente_Miembro         5000 non-null   object 
 13  Descuento_Miembro       5000 non-null   int64  
 14  Promoción_Aplicada      5000 non-null   

* Se identifica 5 entradas
* Se identifican 13 columnas tipo de dato object
* Se identifica   3 columnas tipo de dato float64
* Se identifica   5 columnas tipo de dato int64
* Se identifica 3.630 datos faltantes o nulos y solo 1370 datos validos
  en columna 'Tamaño'
* Se identifica que columna 'Fecha' contiene datos object
* Se identifica que columna 'Hora' contiene datos object,
  se deja en el mismo tipo de dato

In [117]:
df.describe().round(2)

Cantidad  Precio_Unitario  Total_Venta  Descuento_Miembro  \
count   5000.00          5000.00      5000.00            5000.00   
mean       3.01          1714.82      5164.75               3.98   
std        1.41           589.98      3134.05               5.99   
min        1.00           700.01       702.91               0.00   
25%        2.00          1305.76      2753.02               0.00   
50%        3.00          1605.34      4709.99               0.00   
75%        4.00          1955.03      6842.79              10.00   
max        5.00          3498.88     17377.25              15.00   

       Demora_Preparación_Min  Satisfacción_Cliente  Stock_Antes  \
count                 5000.00               5000.00      5000.00   
mean                     8.56                  3.03        30.06   
std                      3.76                  1.40        11.77   
min                      2.00                  1.00        10.00   
25%                      5.30                  2.00        20.00   
50%                      8.60                  3.00        30.00   
75%                     11.80                  4.00        40.00   
max                     15.00                  5.00        50.00   

       Stock_Después  
count        5000.00  
mean           27.05  
std            11.84  
min             5.00  
25%            17.00  
50%            27.00  
75%            37.00  
max            49.00

#### Se deduce de Estadistica que:
* La demora minima en preparacion del pedido esta en el rango de 2 a 15 minutos, con un promedio de 8.56 min. y desviacion standar de 3.76 min.
* La satisfacción del cliente esta en el rango de 1.0 a 5.0, con un promedio de 3.03 y desviacion estandar de 1.40
* La cantidad de Productos comprados esta en el rango de 1 a 5 productos, con un promedio de 3.01 y desviacion standar de 1.41.
* El descuento a los miembros esta en el rango de 0.0 a 15.00 %, con un promedio de 3.98% y desviacion standar de 5.99%.

Se supone que el descuento esta en %, porque no se expresa en el dataset.

#### 4.- Limpieza de Datos

In [118]:
#  Se modifica los nombres de las columnas dejando todas en minusculas y sin espacios.
df.columns = df.columns.str.lower().str.strip()

In [119]:
# se identifican las columnas
df.columns

Index(['id_venta', 'fecha', 'hora', 'sucursal', 'nombre_producto',
       'categoría_producto', 'tamaño', 'cantidad', 'precio_unitario',
       'total_venta', 'método_de_compra', 'método_de_pago', 'cliente_miembro',
       'descuento_miembro', 'promoción_aplicada', 'empleado_atendió', 'turno',
       'demora_preparación_min', 'satisfacción_cliente', 'stock_antes',
       'stock_después'],
      dtype='object')

In [120]:
# Se cambia el tipo de dato de columna 'fecha' a datetime64[ns]

df["fecha"] = pd.to_datetime(df["fecha"])

In [121]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   id_venta                5000 non-null   object        
 1   fecha                   5000 non-null   datetime64[ns]
 2   hora                    5000 non-null   object        
 3   sucursal                5000 non-null   object        
 4   nombre_producto         5000 non-null   object        
 5   categoría_producto      5000 non-null   object        
 6   tamaño                  1370 non-null   object        
 7   cantidad                5000 non-null   int64         
 8   precio_unitario         5000 non-null   float64       
 9   total_venta             5000 non-null   float64       
 10  método_de_compra        5000 non-null   object        
 11  método_de_pago          5000 non-null   object        
 12  cliente_miembro         5000 non-null   object  

In [122]:
# Se revisa si hay duplicados

df.duplicated().sum()

np.int64(0)

 #### Se verifica que no hay entradas duplicadas

 #### Manejo de datos Nulos

In [123]:
# Se identifica los faltantes o nulos en la columna 'tamaño'

df[df["tamaño"].isna()]

id_venta      fecha   hora                       sucursal  \
0     VTA00001 2025-01-04  08:25  San Lorenzo 25, Nueva Córdoba   
1     VTA00002 2025-01-10  14:08               Córdoba Shopping   
2     VTA00003 2025-01-17  15:10               Córdoba Shopping   
3     VTA00004 2025-03-23  07:23           Calle San Lorenzo 47   
4     VTA00005 2025-03-19  18:33               Córdoba Shopping   
...        ...        ...    ...                            ...   
4994  VTA04995 2025-03-26  08:33           Calle San Lorenzo 47   
4995  VTA04996 2025-03-06  08:07              Avenida Colón 608   
4996  VTA04997 2025-01-17  17:46              Avenida Colón 608   
4997  VTA04998 2025-02-06  10:18  San Lorenzo 25, Nueva Córdoba   
4998  VTA04999 2025-03-12  11:10               Córdoba Shopping   

                 nombre_producto categoría_producto tamaño  cantidad  \
0             Termo Reutilizable      Merchandising    NaN         2   
1             Té Verde en Hebras                 Té    NaN         1   
2            Moneda de Chocolate             Snacks    NaN         4   
3                 Espresso Roast     Café en Granos    NaN         5   
4                 Espresso Roast     Café en Granos    NaN         5   
...                          ...                ...    ...       ...   
4994       Té Helado de Hibiscus          Té Helado    NaN         4   
4995  Fosforito de Jamón y Queso         Sándwiches    NaN         2   
4996         Moneda de Chocolate             Snacks    NaN         2   
4997        Croissant de Manteca             Bakery    NaN         2   
4998   Wrap de Pollo y Vegetales         Sándwiches    NaN         5   

      precio_unitario  total_venta  ... método_de_pago cliente_miembro  \
0             3277.87      6555.74  ...            App              Sí   
1             1630.18      1630.18  ...        Tarjeta              Sí   
2              879.24      3516.96  ...        Tarjeta              No   
3             2230.73     11153.65  ...        Tarjeta              Sí   
4             2486.63     12433.15  ...        Tarjeta              No   
...               ...          ...  ...            ...             ...   
4994          1146.33      4585.32  ...        Tarjeta              Sí   
4995          1694.30      3388.60  ...            App              Sí   
4996          1106.72      2213.44  ...        Tarjeta              Sí   
4997          1430.26      2860.52  ...        Tarjeta              No   
4998          1524.23      7621.15  ...        Tarjeta              No   

     descuento_miembro  promoción_aplicada empleado_atendió   turno  \
0                   15               Combo           Camila   Noche   
1                   10             Ninguna        Florencia   Noche   
2                    0             Ninguna           Julián   Tarde   
3                   10             Ninguna           Marcos   Noche   
4                    0               Combo            Sofía   Tarde   
...                ...                 ...              ...     ...   
4994                 0                 2x1        Florencia  Mañana   
4995                 0          Happy Hour           Julián   Noche   
4996                10             Ninguna           Camila  Mañana   
4997                 0             Ninguna           Marcos  Mañana   
4998                 0          Happy Hour        Florencia  Mañana   

     demora_preparación_min  satisfacción_cliente  stock_antes  stock_después  
0                      13.7                     5           23             21  
1                       2.3                     4           20             19  
2                       8.8                     2           23             19  
3                       3.9                     3           19             14  
4                      13.8                     5           41             36  
...                     ...                   ...          ...            ...  
4994                   12.5       

Viendo la data y los valores de las columnas 'categoria_producto' y 'nombre_producto', se supone que los productos que no tienen valor en la columna 'tamaño', son los que no necesitan ser catalogados con un tamaño, como por ejemplo los productos "Moneda de Chocolate" y "Wrap de Pollo y Vegetales	", en el supuesto que sus tamaños son unicos, por lo tanto se cambiara los Nulos por el  valor 'Normal'.  

In [124]:
# Se seleciona todos los nombres de productos que en columna 'tamaño' tienen datos Nulos y con la funcion 'Loc' y se le asigna el valor 'Normal'

df.loc[((df["nombre_producto"] == "Croissant Relleno con Crema de Avellanas") |
        (df["nombre_producto"] == "Moneda de Chocolate") |
         (df["nombre_producto"] == "Termo Reutilizable") |
         (df["nombre_producto"] == "Té Verde en Hebras") |
         (df["nombre_producto"] == "Mix de Frutos Secos") |
         (df["nombre_producto"] == "Té Chai Latte") |
        (df["nombre_producto"] == "Fosforito de Jamón y Queso") |
        (df["nombre_producto"] == "Croissant de Manteca") |
        (df["nombre_producto"] == "Wrap de Pollo y Vegetales") |
        (df["nombre_producto"] == "Cookie con Chips de Chocolate") |
        (df["nombre_producto"] == "Wrap de Pollo y Vegetales") |
        (df["nombre_producto"] == "Té Helado de Hibiscus") |
         (df["nombre_producto"] == "Té Verde") |
         (df["nombre_producto"] == "Bagel Sandwich") |
         (df["nombre_producto"] == "Taza Edición Especial") |
         (df["nombre_producto"] == "Muffin de Arándanos") |
         (df["nombre_producto"] == "Taza Edición Especial")), ["tamaño"]] = "Normal"

In [125]:
df.tail(5)

id_venta      fecha   hora                       sucursal  \
4995  VTA04996 2025-03-06  08:07              Avenida Colón 608   
4996  VTA04997 2025-01-17  17:46              Avenida Colón 608   
4997  VTA04998 2025-02-06  10:18  San Lorenzo 25, Nueva Córdoba   
4998  VTA04999 2025-03-12  11:10               Córdoba Shopping   
4999  VTA05000 2025-03-08  15:54              Avenida Colón 608   

                 nombre_producto      categoría_producto  tamaño  cantidad  \
4995  Fosforito de Jamón y Queso              Sándwiches  Normal         2   
4996         Moneda de Chocolate                  Snacks  Normal         2   
4997        Croissant de Manteca                  Bakery  Normal         2   
4998   Wrap de Pollo y Vegetales              Sándwiches  Normal         5   
4999                Latte Helado  Bebida Espresso Helada   Venti         5   

      precio_unitario  total_venta  ... método_de_pago cliente_miembro  \
4995          1694.30      3388.60  ...            App              Sí   
4996          1106.72      2213.44  ...        Tarjeta              Sí   
4997          1430.26      2860.52  ...        Tarjeta              No   
4998          1524.23      7621.15  ...        Tarjeta              No   
4999          1648.32      8241.60  ...       Efectivo              No   

     descuento_miembro  promoción_aplicada empleado_atendió   turno  \
4995                 0          Happy Hour           Julián   Noche   
4996                10             Ninguna           Camila  Mañana   
4997                 0             Ninguna           Marcos  Mañana   
4998                 0          Happy Hour        Florencia  Mañana   
4999                 0                 2x1        Florencia   Tarde   

     demora_preparación_min  satisfacción_cliente  stock_antes  stock_después  
4995                   10.9                     2           41             39  
4996                    4.7                     4           26             24  
4997                   11.6                     1           30             28  
4998                    6.0                     1           17             12  
4999                   13.7                     5           39             34  

[5 rows x 21 columns]

In [126]:
# Se ve que para el producto de categoria 'Café en Granos', si podria tener distintos tamaños en las distintas sucursales, pero ademas tiene distintos precios.

fil = df[df["categoría_producto"] == "Café en Granos"]
fil[["nombre_producto", "precio_unitario","sucursal"]].value_counts()

nombre_producto  precio_unitario  sucursal                     
House Blend      2999.69          Avenida Colón 608                1
Espresso Roast   2001.34          Córdoba Shopping                 1
                 2002.25          Calle San Lorenzo 47             1
                 2009.57          Calle San Lorenzo 47             1
                 2011.69          Avenida Colón 608                1
                                                                  ..
                 2055.66          San Lorenzo 25, Nueva Córdoba    1
                 2054.81          Avenida Colón 608                1
                 2052.05          Córdoba Shopping                 1
                 2050.33          Avenida Colón 608                1
                 2049.23          Avenida Colón 608                1
Name: count, Length: 459, dtype: int64

In [127]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   id_venta                5000 non-null   object        
 1   fecha                   5000 non-null   datetime64[ns]
 2   hora                    5000 non-null   object        
 3   sucursal                5000 non-null   object        
 4   nombre_producto         5000 non-null   object        
 5   categoría_producto      5000 non-null   object        
 6   tamaño                  4452 non-null   object        
 7   cantidad                5000 non-null   int64         
 8   precio_unitario         5000 non-null   float64       
 9   total_venta             5000 non-null   float64       
 10  método_de_compra        5000 non-null   object        
 11  método_de_pago          5000 non-null   object        
 12  cliente_miembro         5000 non-null   object  

Para la  categoria  "Café en Granos" se hace mas complicado deducir el dato faltante, puesto que tienen distinto precio en distintas sucursales, ademas el  precio unitario puede ser afectado por los distintos descuentos, por ser el cliente miembro o por las distintas promociones. Buscando información los tamaños de cafe en Starbuck son : Corto (8 oz), Alto (12 oz), Grande (16 oz), Venti (20 oz) y Trenta (24 oz). Por lo tanto suponemos un promedio de (16 oz) correpondiente al tamaño Grande. Se llenaraon los campos con el valor 'Grande'.

In [128]:
# Se despliegan los registros con nulos

df[df["tamaño"].isna()]

id_venta      fecha   hora                       sucursal  \
3     VTA00004 2025-03-23  07:23           Calle San Lorenzo 47   
4     VTA00005 2025-03-19  18:33               Córdoba Shopping   
8     VTA00009 2025-02-25  13:53  San Lorenzo 25, Nueva Córdoba   
26    VTA00027 2025-03-10  09:07  San Lorenzo 25, Nueva Córdoba   
27    VTA00028 2025-03-24  19:24           Calle San Lorenzo 47   
...        ...        ...    ...                            ...   
4906  VTA04907 2025-01-13  16:31           Calle San Lorenzo 47   
4951  VTA04952 2025-02-23  09:48              Avenida Colón 608   
4957  VTA04958 2025-02-13  11:45           Calle San Lorenzo 47   
4965  VTA04966 2025-03-29  09:13           Calle San Lorenzo 47   
4987  VTA04988 2025-02-02  15:22               Córdoba Shopping   

     nombre_producto categoría_producto tamaño  cantidad  precio_unitario  \
3     Espresso Roast     Café en Granos    NaN         5          2230.73   
4     Espresso Roast     Café en Granos    NaN         5          2486.63   
8     Espresso Roast     Café en Granos    NaN         4          2519.29   
26       House Blend     Café en Granos    NaN         5          2430.53   
27       House Blend     Café en Granos    NaN         4          2270.69   
...              ...                ...    ...       ...              ...   
4906  Espresso Roast     Café en Granos    NaN         4          2942.20   
4951  Espresso Roast     Café en Granos    NaN         5          2015.97   
4957     House Blend     Café en Granos    NaN         1          2630.41   
4965     House Blend     Café en Granos    NaN         5          2388.04   
4987  Espresso Roast     Café en Granos    NaN         5          2842.83   

      total_venta  ... método_de_pago cliente_miembro descuento_miembro  \
3        11153.65  ...        Tarjeta              Sí                10   
4        12433.15  ...        Tarjeta              No                 0   
8        10077.16  ...            App              No                 0   
26       12152.65  ...       Efectivo              Sí                 0   
27        9082.76  ...       Efectivo              No                 0   
...           ...  ...            ...             ...               ...   
4906     11768.80  ...            App              No                 0   
4951     10079.85  ...            App              No                 0   
4957      2630.41  ...       Efectivo              Sí                 0   
4965     11940.20  ...        Tarjeta              No                 0   
4987     14214.15  ...       Efectivo              Sí                15   

      promoción_aplicada empleado_atendió   turno demora_preparación_min  \
3                Ninguna           Marcos   Noche                    3.9   
4                  Combo            Sofía   Tarde                   13.8   
8                  Combo           Marcos   Noche                    5.0   
26                 Combo            Sofía  Mañana                    9.2   
27                   2x1           Camila  Mañana                    3.9   
...                  ...              ...     ...                    ...   
4906          Happy Hour        Florencia   Tarde                   11.2   
4951               Combo        Florencia   Noche                   10.6   
4957                 2x1           Julián  Mañana                   13.6   
4965               Combo             Luis   Noche                    8.8   
4987             Ninguna        Florencia   Tarde                   13.6   

      satisfacción_cliente  stock_antes  stock_después  
3                        3           19             14  
4                        5           41             36  
8                        1           29             25  
26                       3           28             23  
27                       4           40             36  
...                    ...          ...            ...  
4906                     1           50             46  
4951  

In [129]:
# Se identifica un producto con nulos

df[df["nombre_producto"] == "Espresso Roast"]

id_venta      fecha   hora                       sucursal  \
3     VTA00004 2025-03-23  07:23           Calle San Lorenzo 47   
4     VTA00005 2025-03-19  18:33               Córdoba Shopping   
8     VTA00009 2025-02-25  13:53  San Lorenzo 25, Nueva Córdoba   
74    VTA00075 2025-01-13  20:37  San Lorenzo 25, Nueva Córdoba   
87    VTA00088 2025-02-28  09:16           Calle San Lorenzo 47   
...        ...        ...    ...                            ...   
4847  VTA04848 2025-03-19  07:20              Avenida Colón 608   
4857  VTA04858 2025-02-19  14:23              Avenida Colón 608   
4906  VTA04907 2025-01-13  16:31           Calle San Lorenzo 47   
4951  VTA04952 2025-02-23  09:48              Avenida Colón 608   
4987  VTA04988 2025-02-02  15:22               Córdoba Shopping   

     nombre_producto categoría_producto tamaño  cantidad  precio_unitario  \
3     Espresso Roast     Café en Granos    NaN         5          2230.73   
4     Espresso Roast     Café en Granos    NaN         5          2486.63   
8     Espresso Roast     Café en Granos    NaN         4          2519.29   
74    Espresso Roast     Café en Granos    NaN         4          2504.64   
87    Espresso Roast     Café en Granos    NaN         3          2988.34   
...              ...                ...    ...       ...              ...   
4847  Espresso Roast     Café en Granos    NaN         4          2396.76   
4857  Espresso Roast     Café en Granos    NaN         3          2708.53   
4906  Espresso Roast     Café en Granos    NaN         4          2942.20   
4951  Espresso Roast     Café en Granos    NaN         5          2015.97   
4987  Espresso Roast     Café en Granos    NaN         5          2842.83   

      total_venta  ... método_de_pago cliente_miembro descuento_miembro  \
3        11153.65  ...        Tarjeta              Sí                10   
4        12433.15  ...        Tarjeta              No                 0   
8        10077.16  ...            App              No                 0   
74       10018.56  ...        Tarjeta              Sí                 0   
87        8965.02  ...            App              Sí                10   
...           ...  ...            ...             ...               ...   
4847      9587.04  ...        Tarjeta              No                 0   
4857      8125.59  ...       Efectivo              No                 0   
4906     11768.80  ...            App              No                 0   
4951     10079.85  ...            App              No                 0   
4987     14214.15  ...       Efectivo              Sí                15   

      promoción_aplicada empleado_atendió   turno demora_preparación_min  \
3                Ninguna           Marcos   Noche                    3.9   
4                  Combo            Sofía   Tarde                   13.8   
8                  Combo           Marcos   Noche                    5.0   
74            Happy Hour           Julián  Mañana                    2.8   
87               Ninguna           Camila  Mañana                    2.4   
...                  ...              ...     ...                    ...   
4847                 2x1           Julián   Tarde                    4.1   
4857                 2x1           Julián   Noche                   13.2   
4906          Happy Hour        Florencia   Tarde                   11.2   
4951               Combo        Florencia   Noche                   10.6   
4987             Ninguna        Florencia   Tarde                   13.6   

      satisfacción_cliente  stock_antes  stock_después  
3                        3           19             14  
4                        5           41             36  
8                        1           29             25  
74                       2           37             33  
87                       1           20             17  
...                    ...          ...            ...  
4847                     1           27             23  
4857  

In [130]:
# Se asigna el valor 'Grande' a los nombres mencionados con la funcion 'Loc'

df.loc[((df["nombre_producto"] == "Espresso Roast") |
        (df["nombre_producto"] == "House Blend")), ["tamaño"]] = "Grande"

In [131]:
# Se revisan nuevamente si hay mas nulos

df[df["tamaño"].isna()]

id_venta      fecha   hora                       sucursal  \
105   VTA00106 2025-01-19  10:43              Avenida Colón 608   
279   VTA00280 2025-02-28  07:33           Calle San Lorenzo 47   
284   VTA00285 2025-03-14  07:37               Córdoba Shopping   
352   VTA00353 2025-04-03  20:35  San Lorenzo 25, Nueva Córdoba   
404   VTA00405 2025-03-30  08:46              Avenida Colón 608   
...        ...        ...    ...                            ...   
4672  VTA04673 2025-01-05  14:52               Córdoba Shopping   
4798  VTA04799 2025-03-12  15:40           Calle San Lorenzo 47   
4870  VTA04871 2025-02-02  19:50               Córdoba Shopping   
4889  VTA04890 2025-02-05  11:38               Córdoba Shopping   
4903  VTA04904 2025-01-16  11:44               Córdoba Shopping   

     nombre_producto categoría_producto tamaño  cantidad  precio_unitario  \
105   Budín de Limón             Bakery    NaN         5          1313.82   
279   Budín de Limón             Bakery    NaN         1          1013.74   
284   Budín de Limón             Bakery    NaN         4          1103.83   
352   Budín de Limón             Bakery    NaN         3          1302.41   
404   Budín de Limón             Bakery    NaN         3           992.07   
...              ...                ...    ...       ...              ...   
4672  Budín de Limón             Bakery    NaN         3          1178.21   
4798  Budín de Limón             Bakery    NaN         1           977.26   
4870  Budín de Limón             Bakery    NaN         1          1334.52   
4889  Budín de Limón             Bakery    NaN         2           910.49   
4903  Budín de Limón             Bakery    NaN         5          1318.76   

      total_venta  ... método_de_pago cliente_miembro descuento_miembro  \
105       6569.10  ...       Efectivo              Sí                15   
279       1013.74  ...        Tarjeta              Sí                15   
284       4415.32  ...       Efectivo              Sí                 0   
352       3907.23  ...       Efectivo              No                 0   
404       2976.21  ...            App              Sí                 0   
...           ...  ...            ...             ...               ...   
4672      3534.63  ...        Tarjeta              Sí                15   
4798       977.26  ...       Efectivo              No                 0   
4870      1334.52  ...       Efectivo              No                 0   
4889      1820.98  ...        Tarjeta              No                 0   
4903      6593.80  ...       Efectivo              No                 0   

      promoción_aplicada empleado_atendió   turno demora_preparación_min  \
105              Ninguna             Luis   Noche                    9.8   
279                  2x1           Camila   Tarde                   14.4   
284                  2x1             Luis   Tarde                   12.0   
352                Combo            Sofía   Noche                    4.5   
404           Happy Hour            Sofía   Tarde                   13.0   
...                  ...              ...     ...                    ...   
4672          Happy Hour             Luis  Mañana                    4.4   
4798                 2x1             Luis  Mañana                   13.9   
4870          Happy Hour           Marcos   Tarde                    6.1   
4889             Ninguna           Camila   Tarde                    4.8   
4903                 2x1           Marcos  Mañana                   13.8   

      satisfacción_cliente  stock_antes  stock_después  
105                      5           42             37  
279                      5           39             38  
284                      3           48             44  
352                      5           29             26  
404                      2           31             28  
...                    ...          ...            ...  
4672                     5           31             28  
4798  

In [132]:
# Se vuelve a aplicar el codigo a los productos con nulos faltantes

df.loc[((df["nombre_producto"] == "Budín de Limón")), ["tamaño"]] = "Normal"

In [133]:
# Se vuelve a revisar si quedan registros con valores nulos

df[df["tamaño"].isna()]

Empty DataFrame
Columns: [id_venta, fecha, hora, sucursal, nombre_producto, categoría_producto, tamaño, cantidad, precio_unitario, total_venta, método_de_compra, método_de_pago, cliente_miembro, descuento_miembro, promoción_aplicada, empleado_atendió, turno, demora_preparación_min, satisfacción_cliente, stock_antes, stock_después]
Index: []

[0 rows x 21 columns]

In [134]:
# se revisa la informacion de la data

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   id_venta                5000 non-null   object        
 1   fecha                   5000 non-null   datetime64[ns]
 2   hora                    5000 non-null   object        
 3   sucursal                5000 non-null   object        
 4   nombre_producto         5000 non-null   object        
 5   categoría_producto      5000 non-null   object        
 6   tamaño                  5000 non-null   object        
 7   cantidad                5000 non-null   int64         
 8   precio_unitario         5000 non-null   float64       
 9   total_venta             5000 non-null   float64       
 10  método_de_compra        5000 non-null   object        
 11  método_de_pago          5000 non-null   object        
 12  cliente_miembro         5000 non-null   object  

#### 5.- Transformación de Datos

In [135]:
# Crea nuevas columnas basadas en operaciones con las columnas existentes (por ejemplo, calcular ingresos a partir de ventas y precios).

df["ventas_totales_2"] = df["precio_unitario"]*df["cantidad"]

In [136]:
df.head()

id_venta      fecha   hora                       sucursal  \
0  VTA00001 2025-01-04  08:25  San Lorenzo 25, Nueva Córdoba   
1  VTA00002 2025-01-10  14:08               Córdoba Shopping   
2  VTA00003 2025-01-17  15:10               Córdoba Shopping   
3  VTA00004 2025-03-23  07:23           Calle San Lorenzo 47   
4  VTA00005 2025-03-19  18:33               Córdoba Shopping   

       nombre_producto categoría_producto  tamaño  cantidad  precio_unitario  \
0   Termo Reutilizable      Merchandising  Normal         2          3277.87   
1   Té Verde en Hebras                 Té  Normal         1          1630.18   
2  Moneda de Chocolate             Snacks  Normal         4           879.24   
3       Espresso Roast     Café en Granos  Grande         5          2230.73   
4       Espresso Roast     Café en Granos  Grande         5          2486.63   

   total_venta  ... cliente_miembro descuento_miembro promoción_aplicada  \
0      6555.74  ...              Sí                15              Combo   
1      1630.18  ...              Sí                10            Ninguna   
2      3516.96  ...              No                 0            Ninguna   
3     11153.65  ...              Sí                10            Ninguna   
4     12433.15  ...              No                 0              Combo   

   empleado_atendió  turno demora_preparación_min satisfacción_cliente  \
0            Camila  Noche                   13.7                    5   
1         Florencia  Noche                    2.3                    4   
2            Julián  Tarde                    8.8                    2   
3            Marcos  Noche                    3.9                    3   
4             Sofía  Tarde                   13.8                    5   

   stock_antes  stock_después  ventas_totales_2  
0           23             21           6555.74  
1           20             19           1630.18  
2           23             19           3516.96  
3           19             14          11153.65  
4           41             36          12433.15  

[5 rows x 22 columns]

In [ ]:
# Normaliza o estandariza columnas no es necesario.

#### Clasifica los datos en categorías relevantes.

### Se clasifica categorias relevantes por ejemplo :


In [137]:
# Se clasifica nombres de productos por categoria de productos
df[["categoría_producto", "nombre_producto"]].value_counts()


categoría_producto        nombre_producto                         
Té                        Té Verde en Hebras                          481
Frappuccino               Caramel Frappuccino                         458
Té Helado                 Té Helado de Hibiscus                       415
Té Caliente               Té Verde                                    255
                          Té Chai Latte                               242
Snacks                    Moneda de Chocolate                         238
Café en Granos            Espresso Roast                              236
Merchandising             Taza Edición Especial                       231
Café en Granos            House Blend                                 223
Merchandising             Termo Reutilizable                          211
Snacks                    Mix de Frutos Secos                         207
Sándwiches                Fosforito de Jamón y Queso                  156
                          Wrap de Pollo y Vegetales                   148
                          Bagel Sandwich                              113
Bakery                    Muffin de Arándanos                         105
                          Cookie con Chips de Chocolate               105
Bebida Espresso Helada    Skinny Vainilla Latte Helado                102
                          Latte Helado                                 94
Bakery                    Budín de Limón                               89
                          Croissant de Manteca                         89
                          Croissant Relleno con Crema de Avellanas     86
Bebida Espresso Helada    Americano Helado                             86
                          Cold Brew                                    84
                          Vainilla Latte Helado                        81
Bebida Espresso Caliente  Mocha Blanco                                 49
                          Skinny Vainilla Latte                        43
                          Flat White                                   41
                          Café del Día                                 40
                          Latte Macchiato                              40
                          Cappuccino                                   40
                          Americano                                    40
                          Latte                                        39
                          Mocha                                        39
                          Café del Día con Leche                       33
                          Vainilla Latte                               31
                          Skinny Caramel Macchiato                     30
Name: count, dtype: int64

Se deduce de los datos que la categoria mas vendida es el 'Té'.

In [138]:
# Se clasifica atenciones de empleados por turno y sucursal

df[["sucursal", "turno", "nombre_producto"]].value_counts()


sucursal                       turno   nombre_producto       
Calle San Lorenzo 47           Noche   Caramel Frappuccino       51
                               Tarde   Té Verde en Hebras        50
                               Mañana  Té Verde en Hebras        46
San Lorenzo 25, Nueva Córdoba  Noche   Té Verde en Hebras        45
                               Tarde   Caramel Frappuccino       45
                                                                 ..
Avenida Colón 608              Noche   Vainilla Latte             1
San Lorenzo 25, Nueva Córdoba  Noche   Flat White                 1
Avenida Colón 608              Noche   Americano                  1
San Lorenzo 25, Nueva Córdoba  Tarde   Café del Día con Leche     1
                                       Vainilla Latte             1
Name: count, Length: 431, dtype: int64

####  6.- Análisis de Datos

* Realiza agrupaciones de datos utilizando groupby para obtener insights específicos (por ejemplo, ventas por producto, ventas por región, etc.).

In [139]:
 # Se muestra las sucursales

 df["sucursal"].value_counts().sort_values()

sucursal
Avenida Colón 608                1229
Córdoba Shopping                 1229
San Lorenzo 25, Nueva Córdoba    1267
Calle San Lorenzo 47             1275
Name: count, dtype: int64

* Se deduce que hay 4 sucursales
* Se deduce sucursal con mayor cantidad de resgistros es 'Avenida Colón 608'

In [140]:
# Ventas por turno y sucursal

df.groupby(["sucursal", "turno"])["total_venta"].sum().sort_values(ascending=False)

sucursal                       turno 
Avenida Colón 608              Noche     2266190.66
Calle San Lorenzo 47           Tarde     2252284.54
Avenida Colón 608              Mañana    2231829.01
Córdoba Shopping               Tarde     2226639.07
San Lorenzo 25, Nueva Córdoba  Noche     2222377.50
Calle San Lorenzo 47           Noche     2172024.18
San Lorenzo 25, Nueva Córdoba  Tarde     2160987.70
                               Mañana    2136661.15
Calle San Lorenzo 47           Mañana    2099576.28
Córdoba Shopping               Mañana    2047135.42
Avenida Colón 608              Tarde     2016867.65
Córdoba Shopping               Noche     1991200.92
Name: total_venta, dtype: float64

* Se deduce en la sucursal 'Avenida Colón 608' en turno 'Noche' tiene la mayores ventas

In [141]:
 # Calcular total de clientes que son miembros con descuento miembro y los que no son miembros.

df.groupby(["cliente_miembro"])["descuento_miembro"].size()

cliente_miembro
No    2550
Sí    2450
Name: descuento_miembro, dtype: int64

* Se deduce que la cantidad de clientes miembros con descuento de miembro son 2450
* Se deduce que la cantidad de clientes sin membresia ni descuento de miembro son 2550

In [142]:
# Se calcula ventas promedio por sucursal.

vta_promedio = df.groupby(df["sucursal"])["total_venta"].mean().round(2).sort_values(ascending=False)
print(vta_promedio)

sucursal
Avenida Colón 608                5300.97
San Lorenzo 25, Nueva Córdoba    5146.04
Calle San Lorenzo 47             5116.77
Córdoba Shopping                 5097.62
Name: total_venta, dtype: float64


Se deduce que la sucursal con el mayor promedio de ventas es la sucursal 'Avenida Colón 608' con 5300.97

In [143]:
# Se calcula promedio de unidades vendidas de productos, por sucursal y categoria de producto.

df.groupby(["sucursal", "categoría_producto", "nombre_producto"])["cantidad"].mean().round(2).sort_values(ascending=False)

sucursal                       categoría_producto        nombre_producto       
Calle San Lorenzo 47           Bebida Espresso Helada    Cold Brew                 3.67
Avenida Colón 608              Bakery                    Budín de Limón            3.62
Córdoba Shopping               Bebida Espresso Caliente  Café del Día              3.60
Calle San Lorenzo 47           Bebida Espresso Caliente  Cappuccino                3.57
Córdoba Shopping               Bebida Espresso Caliente  Latte                     3.50
                                                                                   ... 
Avenida Colón 608              Bebida Espresso Caliente  Americano                 2.12
San Lorenzo 25, Nueva Córdoba  Bebida Espresso Caliente  Cappuccino                2.08
Calle San Lorenzo 47           Bebida Espresso Caliente  Café del Día con Leche    2.00
San Lorenzo 25, Nueva Córdoba  Bebida Espresso Caliente  Latte                     2.00
Avenida Colón 608              Bebida Espresso Caliente  Skinny Vainilla Latte     1.89
Name: cantidad, Length: 144, dtype: float64

Se deduce que el promedio mas alto de unidades vendidas fue el productos 'Cold Brew' en la sucursal de 'Calle San Lorenzo 47' con valor de 3.67

In [144]:
# Se Aplica el metodo apply para redondear a entero los valores de la columna 'demora_preparación_min'

df["demora_preparación_min"] = df["demora_preparación_min"].apply(lambda x: round(x))

In [146]:
df.head()

id_venta      fecha   hora                       sucursal  \
0  VTA00001 2025-01-04  08:25  San Lorenzo 25, Nueva Córdoba   
1  VTA00002 2025-01-10  14:08               Córdoba Shopping   
2  VTA00003 2025-01-17  15:10               Córdoba Shopping   
3  VTA00004 2025-03-23  07:23           Calle San Lorenzo 47   
4  VTA00005 2025-03-19  18:33               Córdoba Shopping   

       nombre_producto categoría_producto  tamaño  cantidad  precio_unitario  \
0   Termo Reutilizable      Merchandising  Normal         2          3277.87   
1   Té Verde en Hebras                 Té  Normal         1          1630.18   
2  Moneda de Chocolate             Snacks  Normal         4           879.24   
3       Espresso Roast     Café en Granos  Grande         5          2230.73   
4       Espresso Roast     Café en Granos  Grande         5          2486.63   

   total_venta  ... cliente_miembro descuento_miembro promoción_aplicada  \
0      6555.74  ...              Sí                15              Combo   
1      1630.18  ...              Sí                10            Ninguna   
2      3516.96  ...              No                 0            Ninguna   
3     11153.65  ...              Sí                10            Ninguna   
4     12433.15  ...              No                 0              Combo   

   empleado_atendió  turno demora_preparación_min satisfacción_cliente  \
0            Camila  Noche                     14                    5   
1         Florencia  Noche                      2                    4   
2            Julián  Tarde                      9                    2   
3            Marcos  Noche                      4                    3   
4             Sofía  Tarde                     14                    5   

   stock_antes  stock_después  ventas_totales_2  
0           23             21           6555.74  
1           20             19           1630.18  
2           23             19           3516.96  
3           19             14          11153.65  
4           41             36          12433.15  

[5 rows x 22 columns]